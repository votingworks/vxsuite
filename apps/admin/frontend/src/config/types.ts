import {
  BallotStyleId,
  ContestId,
  PrecinctId,
  PromiseOr,
} from '@votingworks/types';
import type { ManualResultsVotingMethod } from '@votingworks/admin-backend';

// Events
export type InputEventFunction = (
  event: React.FormEvent<HTMLInputElement>
) => PromiseOr<void>;
export type TextareaEventFunction = (
  event: React.FormEvent<HTMLTextAreaElement>
) => PromiseOr<void>;

// Router Props
export interface ManualDataEntryScreenProps {
  precinctId: PrecinctId;
  ballotStyleId: BallotStyleId;
  votingMethod: ManualResultsVotingMethod;
}
export interface SmartcardsScreenProps {
  smartcardType: string;
}
export interface WriteInsAdjudicationScreenProps {
  contestId: ContestId;
}

export enum ResultsFileType {
  CastVoteRecord = 'cvr',
  All = 'all',
}

export interface CastVoteRecordFilePreprocessedData {
  readonly name: string;
  readonly path: string;
  readonly cvrCount: number;
  readonly scannerIds: readonly string[];
  readonly exportTimestamp: Date;
  readonly isTestModeResults: boolean;
}

export type Iso8601Timestamp = string;
