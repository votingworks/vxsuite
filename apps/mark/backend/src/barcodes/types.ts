export interface ScanEvent {
  data: Uint8Array;
}
