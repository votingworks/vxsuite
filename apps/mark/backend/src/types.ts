import { PollsState, PrecinctSelection } from '@votingworks/types';
import { PrintBallotProps as BackendPrintBallotProps } from './util/print_ballot';

export interface MachineConfig {
  machineId: string;
  codeVersion: string;
  screenOrientation: ScreenOrientation;
}

export interface ElectionState {
  precinctSelection?: PrecinctSelection;
  ballotsPrintedCount: number;
  isTestMode: boolean;
  pollsState: PollsState;
}

export type ScreenOrientation = 'portrait' | 'landscape';

export type PrintBallotProps = Omit<
  BackendPrintBallotProps,
  'store' | 'printer'
>;
