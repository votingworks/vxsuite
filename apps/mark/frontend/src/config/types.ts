import {
  BallotStyleId,
  ElectionDefinition,
  PrecinctId,
  VotesDict,
} from '@votingworks/types';
import type { MachineConfig } from '@votingworks/mark-backend';
import {
  ContestsWithMsEitherNeither,
  UpdateVoteFunction,
} from '@votingworks/mark-flow-ui';

// Ballot
export interface BallotContextInterface {
  machineConfig: MachineConfig;
  ballotStyleId?: BallotStyleId;
  contests: ContestsWithMsEitherNeither;
  readonly electionDefinition?: ElectionDefinition;
  generateBallotId: () => string;
  isCardlessVoter: boolean;
  isLiveMode: boolean;
  endVoterSession: () => Promise<void>;
  precinctId?: PrecinctId;
  resetBallot: (showPostVotingInstructions?: boolean) => void;
  updateTally: () => void;
  updateVote: UpdateVoteFunction;
  votes: VotesDict;
}
