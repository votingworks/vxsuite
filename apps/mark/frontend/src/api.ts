import type { Api } from '@votingworks/mark-backend';
import React from 'react';
import * as grout from '@votingworks/grout';
import {
  QueryClient,
  QueryKey,
  useMutation,
  useQuery,
  useQueryClient,
} from '@tanstack/react-query';
import {
  AUTH_STATUS_POLLING_INTERVAL_MS,
  QUERY_CLIENT_DEFAULT_OPTIONS,
} from '@votingworks/ui';

export type ApiClient = grout.Client<Api>;

export function createApiClient(): ApiClient {
  return grout.createClient<Api>({ baseUrl: '/api' });
}

export const ApiClientContext = React.createContext<ApiClient | undefined>(
  undefined
);

export function useApiClient(): ApiClient {
  const apiClient = React.useContext(ApiClientContext);
  if (!apiClient) {
    throw new Error('ApiClientContext.Provider not found');
  }
  return apiClient;
}

export function createQueryClient(): QueryClient {
  return new QueryClient({ defaultOptions: QUERY_CLIENT_DEFAULT_OPTIONS });
}

export const getMachineConfig = {
  queryKey(): QueryKey {
    return ['getMachineConfig'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getMachineConfig());
  },
} as const;

export const getElectionDefinition = {
  queryKey(): QueryKey {
    return ['getElectionDefinition'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getElectionDefinition());
  },
} as const;

/* istanbul ignore next */
export const getSystemSettings = {
  queryKey(): QueryKey {
    return ['getSystemSettings'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getSystemSettings());
  },
} as const;

export const getAuthStatus = {
  queryKey(): QueryKey {
    return ['getAuthStatus'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getAuthStatus(), {
      refetchInterval: AUTH_STATUS_POLLING_INTERVAL_MS,
    });
  },
} as const;

export const checkPin = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.checkPin, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

/* istanbul ignore next */
export const logOut = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.logOut, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

/* istanbul ignore next */
export const updateSessionExpiry = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.updateSessionExpiry, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

export const startCardlessVoterSession = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.startCardlessVoterSession, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

export const endCardlessVoterSession = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.endCardlessVoterSession, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

export const configureBallotPackageFromUsb = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(() => apiClient.configureBallotPackageFromUsb(), {
      async onSuccess() {
        await queryClient.invalidateQueries(getElectionDefinition.queryKey());
        await queryClient.invalidateQueries(getSystemSettings.queryKey());
      },
    });
  },
} as const;

export const unconfigureMachine = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(() => apiClient.unconfigureMachine(), {
      async onSuccess() {
        await queryClient.invalidateQueries(getElectionDefinition.queryKey());
        await queryClient.invalidateQueries(getSystemSettings.queryKey());
      },
    });
  },
} as const;
