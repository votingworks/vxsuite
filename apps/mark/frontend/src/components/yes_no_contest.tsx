import React, {
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  YesNoVote,
  OptionalYesNoVote,
  YesNoContest as YesNoContestInterface,
  YesOrNo,
  getContestDistrictName,
} from '@votingworks/types';
import {
  Button,
  ContestChoiceButton,
  DisplayTextForYesOrNo,
  Main,
  Modal,
  Prose,
  TextWithLineBreaks,
} from '@votingworks/ui';

import { getSingleYesNoVote } from '@votingworks/utils';
import { assert, Optional } from '@votingworks/basics';
import { ScrollDirections, UpdateVoteFunction } from '../config/types';

import { BallotContext } from '../contexts/ballot_context';

import { FONT_SIZES } from '../config/globals';

import {
  ContentHeader,
  ContestFooter,
  DistrictName,
  VariableContentContainer,
  ScrollControls,
  ScrollContainer,
  ScrollableContentWrapper,
  ChoicesGrid,
  ContestDescription,
} from './contest_screen_layout';

interface Props {
  contest: YesNoContestInterface;
  vote: OptionalYesNoVote;
  updateVote: UpdateVoteFunction;
}

export function YesNoContest({
  contest,
  vote,
  updateVote,
}: Props): JSX.Element {
  const { userSettings, electionDefinition } = useContext(BallotContext);
  assert(electionDefinition);
  const { election } = electionDefinition;
  const districtName = getContestDistrictName(election, contest);
  const scrollContainer = useRef<HTMLDivElement>(null);

  const [isScrollable, setIsScrollable] = useState(false);
  const [isScrollAtBottom, setIsScrollAtBottom] = useState(true);
  const [isScrollAtTop, setIsScrollAtTop] = useState(true);
  const [overvoteSelection, setOvervoteSelection] =
    useState<Optional<YesOrNo>>();
  const [deselectedVote, setDeselectedVote] = useState('');

  useEffect(() => {
    if (deselectedVote !== '') {
      const timer = setTimeout(() => setDeselectedVote(''), 100);
      return () => clearTimeout(timer);
    }
  }, [deselectedVote]);

  const updateContestChoicesScrollStates = useCallback(() => {
    const target = scrollContainer.current;
    /* istanbul ignore next - `target` should always exist, but sometimes it doesn't. Don't know how to create this condition in testing.  */
    if (!target) {
      return;
    }
    const targetMinHeight = FONT_SIZES[userSettings.textSize] * 8; // magic number: room for buttons + spacing
    const windowsScrollTopOffsetMagicNumber = 1; // Windows Chrome is often 1px when using scroll buttons.
    const windowsScrollTop = Math.ceil(target.scrollTop); // Windows Chrome scrolls to sub-pixel values.
    setIsScrollable(
      /* istanbul ignore next: Tested by Cypress */
      target.scrollHeight > target.offsetHeight &&
        /* istanbul ignore next: Tested by Cypress */
        target.offsetHeight > targetMinHeight
    );
    setIsScrollAtBottom(
      windowsScrollTop +
        target.offsetHeight +
        windowsScrollTopOffsetMagicNumber >= // Windows Chrome "gte" check.
        target.scrollHeight
    );
    setIsScrollAtTop(target.scrollTop === 0);
  }, [scrollContainer, userSettings.textSize]);

  const voteLength = vote?.length;
  useEffect(() => {
    updateContestChoicesScrollStates();
    window.addEventListener('resize', updateContestChoicesScrollStates);
    return () => {
      window.removeEventListener('resize', updateContestChoicesScrollStates);
    };
  }, [voteLength, updateContestChoicesScrollStates]);

  function handleUpdateSelection(newVote: YesOrNo) {
    if ((vote as string[] | undefined)?.includes(newVote)) {
      updateVote(contest.id, undefined);
      setDeselectedVote(newVote);
    } else {
      updateVote(contest.id, [newVote] as YesNoVote);
    }
  }

  function handleChangeVoteAlert(newValue: YesOrNo) {
    setOvervoteSelection(newValue);
  }

  /* istanbul ignore next: Tested by Cypress */
  function scrollContestChoices(direction: ScrollDirections) {
    const sc = scrollContainer.current;
    assert(sc);
    const currentScrollTop = sc.scrollTop;
    const { offsetHeight } = sc;
    const { scrollHeight } = sc;
    const idealScrollDistance = Math.round(offsetHeight * 0.75);
    const maxScrollableDownDistance =
      scrollHeight - offsetHeight - currentScrollTop;
    const maxScrollTop =
      direction === 'down'
        ? currentScrollTop + maxScrollableDownDistance
        : currentScrollTop;
    const idealScrollTop =
      direction === 'down'
        ? currentScrollTop + idealScrollDistance
        : currentScrollTop - idealScrollDistance;
    const top = idealScrollTop > maxScrollTop ? maxScrollTop : idealScrollTop;
    sc.scrollTo({ behavior: 'smooth', left: 0, top });
  }

  function closeOvervoteAlert() {
    setOvervoteSelection(undefined);
  }

  return (
    <React.Fragment>
      <Main flexColumn>
        <ContentHeader id="contest-header">
          <Prose id="audiofocus">
            <h1 aria-label={`${districtName} ${contest.title}.`}>
              <DistrictName>{districtName}</DistrictName>
              {contest.title}
            </h1>
            <p>
              Vote <strong>Yes</strong> or <strong>No</strong>.
              <span className="screen-reader-only">
                {contest.description}
                To navigate through the contest choices, use the down button. To
                move to the next contest, use the right button.
              </span>
            </p>
          </Prose>
        </ContentHeader>
        <VariableContentContainer
          showTopShadow={!isScrollAtTop}
          showBottomShadow={!isScrollAtBottom}
        >
          <ScrollContainer
            ref={scrollContainer}
            onScroll={updateContestChoicesScrollStates}
          >
            <ScrollableContentWrapper isScrollable={isScrollable}>
              <ContestDescription>
                <Prose>
                  <TextWithLineBreaks text={contest.description} />
                </Prose>
              </ContestDescription>
            </ScrollableContentWrapper>
          </ScrollContainer>
          {
            /* istanbul ignore next: Tested by Cypress */ isScrollable && (
              <ScrollControls aria-hidden>
                <Button
                  className="scroll-up"
                  large
                  variant="primary"
                  aria-hidden
                  value="up"
                  disabled={isScrollAtTop}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
                <Button
                  className="scroll-down"
                  large
                  variant="primary"
                  aria-hidden
                  value="down"
                  disabled={isScrollAtBottom}
                  onPress={scrollContestChoices}
                >
                  <span>See More</span>
                </Button>
              </ScrollControls>
            )
          }
        </VariableContentContainer>
        <ContestFooter>
          <ChoicesGrid data-testid="contest-choices">
            {[
              { label: 'Yes', vote: 'yes' } as const,
              { label: 'No', vote: 'no' } as const,
            ].map((answer) => {
              const isChecked = getSingleYesNoVote(vote) === answer.vote;
              const isDisabled = !isChecked && !!vote;
              function handleDisabledClick() {
                handleChangeVoteAlert(answer.vote);
              }
              let prefixAudioText = '';
              if (isChecked) {
                prefixAudioText = 'Selected,';
              } else if (deselectedVote === answer.vote) {
                prefixAudioText = 'Deselected,';
              }
              return (
                <ContestChoiceButton
                  key={answer.vote}
                  choice={answer.vote}
                  isSelected={isChecked}
                  onPress={
                    isDisabled ? handleDisabledClick : handleUpdateSelection
                  }
                  ariaLabel={`${prefixAudioText} ${answer.label} on ${contest.title}`}
                  label={answer.label}
                />
              );
            })}
          </ChoicesGrid>
        </ContestFooter>
      </Main>
      {overvoteSelection && (
        <Modal
          centerContent
          content={
            <Prose>
              {overvoteSelection && (
                <p id="modalaudiofocus">
                  Do you want to change your vote to{' '}
                  <strong>{DisplayTextForYesOrNo[overvoteSelection]}</strong>?
                  To change your vote, first unselect your vote for{' '}
                  <strong>
                    {
                      {
                        no: DisplayTextForYesOrNo.yes,
                        yes: DisplayTextForYesOrNo.no,
                      }[overvoteSelection]
                    }
                  </strong>
                  .
                </p>
              )}
            </Prose>
          }
          actions={
            <Button variant="primary" autoFocus onPress={closeOvervoteAlert}>
              Okay
            </Button>
          }
        />
      )}
    </React.Fragment>
  );
}
