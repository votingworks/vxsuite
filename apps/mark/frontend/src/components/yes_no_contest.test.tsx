import React from 'react';
import { YesNoContest as YesNoContestInterface } from '@votingworks/types';

import {
  electionMinimalExhaustiveSample,
  electionMinimalExhaustiveSampleDefinition,
} from '@votingworks/fixtures';
import { QueryClientProvider } from '@tanstack/react-query';
import { fireEvent, screen, within } from '../../test/react_testing_library';
import { render as renderWithBallotContext } from '../../test/test_utils';
import { YesNoContest } from './yes_no_contest';
import { ApiClientContext, createQueryClient } from '../api';
import { ApiMock, createApiMock } from '../../test/helpers/mock_api_client';

let apiMock: ApiMock;

const contest = electionMinimalExhaustiveSample.contests.find(
  (c) => c.id === 'fishing' && c.type === 'yesno'
) as YesNoContestInterface;

function renderComponent(
  children: Parameters<typeof renderWithBallotContext>[0],
  renderOptions: Parameters<typeof renderWithBallotContext>[1] = {}
) {
  apiMock.expectGetElectionDefinition(
    renderOptions.electionDefinition ??
      electionMinimalExhaustiveSampleDefinition
  );
  return renderWithBallotContext(
    <ApiClientContext.Provider value={apiMock.mockApiClient}>
      <QueryClientProvider client={createQueryClient()}>
        {children}
      </QueryClientProvider>
    </ApiClientContext.Provider>,
    renderOptions
  );
}

beforeEach(() => {
  apiMock = createApiMock();
});

afterEach(() => {
  apiMock.mockApiClient.assertComplete();
});

describe('supports yes/no contest', () => {
  it('allows voting for both yes and no', async () => {
    const updateVote = jest.fn();
    const { container } = renderComponent(
      <YesNoContest
        contest={contest}
        vote={undefined}
        updateVote={updateVote}
      />,
      {
        electionDefinition: electionMinimalExhaustiveSampleDefinition,
      }
    );

    const contestChoices = await screen.findByTestId('contest-choices');
    expect(container).toMatchSnapshot();

    fireEvent.click(
      (await within(contestChoices).findByText('Yes')).closest('button')!
    );
    expect(updateVote).toHaveBeenCalledTimes(1);

    fireEvent.click(
      (await within(contestChoices).findByText('No')).closest('button')!
    );
    expect(updateVote).toHaveBeenCalledTimes(2);
  });

  it('displays warning when attempting to change vote', async () => {
    const updateVote = jest.fn();
    const { container } = renderComponent(
      <YesNoContest contest={contest} vote={['yes']} updateVote={updateVote} />,
      {
        electionDefinition: electionMinimalExhaustiveSampleDefinition,
      }
    );

    const contestChoices = await screen.findByTestId('contest-choices');
    expect(container).toMatchSnapshot();

    fireEvent.click(within(contestChoices).getByText('No').closest('button')!);
    expect(
      screen.getAllByText(
        (_, element) =>
          element?.textContent ===
          'Do you want to change your vote to No? To change your vote, first unselect your vote for Yes.'
      )
    ).toBeTruthy();
    fireEvent.click(screen.getByText('Okay'));
  });
});
