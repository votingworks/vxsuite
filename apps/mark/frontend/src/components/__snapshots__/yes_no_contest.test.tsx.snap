// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports yes/no contest allows voting for both yes and no 1`] = `
.c0 {
  margin: 0 auto;
  width: 100%;
  padding: 1rem 30px 0.5rem;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  padding: 30px 20px 0;
}

.c1 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  overflow: auto;
}

.c2::before,
.c2::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c2::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c2::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c4 {
  margin: 0 auto;
  width: 100%;
  padding: 0.25rem 20px 0;
}

.c5 {
  padding: 0 10px;
}

.c7 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 20px;
  margin-bottom: 40px;
}

<div>
  <main
    class="sc-ZOsLE fjQwTk"
  >
    <div
      class="c0"
      id="contest-header"
    >
      <div
        class="sc-fotPbf iwtRsQ"
        id="audiofocus"
      >
        <h1
          aria-label="District 1 Ballot Measure 3."
          class="sc-eCImvq cmseZU"
        >
          <div
            class="c1"
          >
            District 1
          </div>
          Ballot Measure 3
        </h1>
        <p
          class="sc-dkPtyc XGKAk"
        >
          Vote 
          <strong>
            Yes
          </strong>
           or 
          <strong>
            No
          </strong>
          .
          <span
            class="screen-reader-only"
          >
            Should fishing be banned in all city owned lakes and rivers?
            To navigate through the contest choices, use the down button. To move to the next contest, use the right button.
          </span>
        </p>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="sc-fotPbf iwtRsQ"
            >
              <p
                class="sc-bqiQRQ bIFmdQ"
              >
                <span>
                  Should fishing be banned in all city owned lakes and rivers?
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        aria-multiselectable="true"
        class="c7"
        data-testid="contest-choices"
        role="listbox"
      >
        <button
          aria-label=" Yes on Ballot Measure 3"
          aria-selected="false"
          class="sc-iJKOzS ebzhGa sc-jcFkyM itMNA-D"
          role="option"
          type="button"
        >
          <span
            class="sc-ezbkgU gcJsOl"
          >
            <span
              class="sc-caiKgP SOLan"
            >
              <span
                class="sc-iUKrWq hYcNmD"
              >
                <span
                  class="sc-cidCJl bYFEFC"
                >
                  <svg
                    aria-hidden="true"
                    class="sc-iwjezw daAwbp"
                    height="100"
                    role="img"
                    viewBox="0 0 100 100"
                    width="100"
                  >
                    <path
                      d="M89.7038 10.1045C88.2094 8.40006 85.759 8.40006 84.2646 10.1045L39.0198 61.5065L15.719 34.8471C14.2245 33.1364 11.7906 33.1364 10.2852 34.8471L2.12082 44.1186C0.626395 45.8105 0.626395 48.5951 2.12082 50.2996L36.2782 89.3708C37.7727 91.0628 40.2066 91.0628 41.7175 89.3708L97.8627 25.5632C99.3791 23.8587 99.3791 21.0679 97.8627 19.3572L89.7038 10.1045Z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="sc-iAKVOt gyainp"
              >
                <p
                  class="sc-dkPtyc XGKAk sc-efQUeY hTUHyO"
                >
                  Yes
                </p>
              </span>
            </span>
          </span>
        </button>
        <button
          aria-label=" No on Ballot Measure 3"
          aria-selected="false"
          class="sc-iJKOzS ebzhGa sc-jcFkyM itMNA-D"
          role="option"
          type="button"
        >
          <span
            class="sc-ezbkgU gcJsOl"
          >
            <span
              class="sc-caiKgP SOLan"
            >
              <span
                class="sc-iUKrWq hYcNmD"
              >
                <span
                  class="sc-cidCJl bYFEFC"
                >
                  <svg
                    aria-hidden="true"
                    class="sc-iwjezw daAwbp"
                    height="100"
                    role="img"
                    viewBox="0 0 100 100"
                    width="100"
                  >
                    <path
                      d="M89.7038 10.1045C88.2094 8.40006 85.759 8.40006 84.2646 10.1045L39.0198 61.5065L15.719 34.8471C14.2245 33.1364 11.7906 33.1364 10.2852 34.8471L2.12082 44.1186C0.626395 45.8105 0.626395 48.5951 2.12082 50.2996L36.2782 89.3708C37.7727 91.0628 40.2066 91.0628 41.7175 89.3708L97.8627 25.5632C99.3791 23.8587 99.3791 21.0679 97.8627 19.3572L89.7038 10.1045Z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="sc-iAKVOt gyainp"
              >
                <p
                  class="sc-dkPtyc XGKAk sc-efQUeY hTUHyO"
                >
                  No
                </p>
              </span>
            </span>
          </span>
        </button>
      </div>
    </div>
  </main>
</div>
`;

exports[`supports yes/no contest displays warning when attempting to change vote 1`] = `
.c0 {
  margin: 0 auto;
  width: 100%;
  padding: 1rem 30px 0.5rem;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  padding: 30px 20px 0;
}

.c1 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  overflow: auto;
}

.c2::before,
.c2::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c2::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c2::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c4 {
  margin: 0 auto;
  width: 100%;
  padding: 0.25rem 20px 0;
}

.c5 {
  padding: 0 10px;
}

.c7 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 20px;
  margin-bottom: 40px;
}

<div>
  <main
    class="sc-ZOsLE fjQwTk"
  >
    <div
      class="c0"
      id="contest-header"
    >
      <div
        class="sc-fotPbf iwtRsQ"
        id="audiofocus"
      >
        <h1
          aria-label="District 1 Ballot Measure 3."
          class="sc-eCImvq cmseZU"
        >
          <div
            class="c1"
          >
            District 1
          </div>
          Ballot Measure 3
        </h1>
        <p
          class="sc-dkPtyc XGKAk"
        >
          Vote 
          <strong>
            Yes
          </strong>
           or 
          <strong>
            No
          </strong>
          .
          <span
            class="screen-reader-only"
          >
            Should fishing be banned in all city owned lakes and rivers?
            To navigate through the contest choices, use the down button. To move to the next contest, use the right button.
          </span>
        </p>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="sc-fotPbf iwtRsQ"
            >
              <p
                class="sc-bqiQRQ bIFmdQ"
              >
                <span>
                  Should fishing be banned in all city owned lakes and rivers?
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        aria-multiselectable="true"
        class="c7"
        data-testid="contest-choices"
        role="listbox"
      >
        <button
          aria-label="Selected, Yes on Ballot Measure 3"
          aria-selected="true"
          class="sc-iJKOzS jNwvxi sc-jcFkyM bpoIao"
          role="option"
          type="button"
        >
          <span
            class="sc-ezbkgU gcJsOl"
          >
            <span
              class="sc-caiKgP SOLan"
            >
              <span
                class="sc-iUKrWq hYcNmD"
              >
                <span
                  class="sc-cidCJl bdaONb"
                >
                  <svg
                    aria-hidden="true"
                    class="sc-iwjezw daAwbp"
                    height="100"
                    role="img"
                    viewBox="0 0 100 100"
                    width="100"
                  >
                    <path
                      d="M89.7038 10.1045C88.2094 8.40006 85.759 8.40006 84.2646 10.1045L39.0198 61.5065L15.719 34.8471C14.2245 33.1364 11.7906 33.1364 10.2852 34.8471L2.12082 44.1186C0.626395 45.8105 0.626395 48.5951 2.12082 50.2996L36.2782 89.3708C37.7727 91.0628 40.2066 91.0628 41.7175 89.3708L97.8627 25.5632C99.3791 23.8587 99.3791 21.0679 97.8627 19.3572L89.7038 10.1045Z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="sc-iAKVOt gyainp"
              >
                <p
                  class="sc-dkPtyc XGKAk sc-efQUeY hTUHyO"
                >
                  Yes
                </p>
              </span>
            </span>
          </span>
        </button>
        <button
          aria-label=" No on Ballot Measure 3"
          aria-selected="false"
          class="sc-iJKOzS ebzhGa sc-jcFkyM itMNA-D"
          role="option"
          type="button"
        >
          <span
            class="sc-ezbkgU gcJsOl"
          >
            <span
              class="sc-caiKgP SOLan"
            >
              <span
                class="sc-iUKrWq hYcNmD"
              >
                <span
                  class="sc-cidCJl bYFEFC"
                >
                  <svg
                    aria-hidden="true"
                    class="sc-iwjezw daAwbp"
                    height="100"
                    role="img"
                    viewBox="0 0 100 100"
                    width="100"
                  >
                    <path
                      d="M89.7038 10.1045C88.2094 8.40006 85.759 8.40006 84.2646 10.1045L39.0198 61.5065L15.719 34.8471C14.2245 33.1364 11.7906 33.1364 10.2852 34.8471L2.12082 44.1186C0.626395 45.8105 0.626395 48.5951 2.12082 50.2996L36.2782 89.3708C37.7727 91.0628 40.2066 91.0628 41.7175 89.3708L97.8627 25.5632C99.3791 23.8587 99.3791 21.0679 97.8627 19.3572L89.7038 10.1045Z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="sc-iAKVOt gyainp"
              >
                <p
                  class="sc-dkPtyc XGKAk sc-efQUeY hTUHyO"
                >
                  No
                </p>
              </span>
            </span>
          </span>
        </button>
      </div>
    </div>
  </main>
</div>
`;
